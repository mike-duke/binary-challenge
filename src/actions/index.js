export const addRelevantArticles = (articles) => ({
  type: 'ADD_RELEVANT_ARTICLES',
  articles
})

export const addCurrentArticles = (articles) => ({
  type: 'ADD_CURRENT_ARTICLES',
  articles
})

export const addTopic = (topic) => ({
  type: 'ADD_TOPIC',
  topic
})