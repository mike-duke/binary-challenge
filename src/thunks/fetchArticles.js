import { addArticles } from '../actions';

export const fetchArticles = (url) => {
  return async (dispatch) => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        console.log(response.statusText);
      }
      const result = await response.json();
      console.log(result.articles);
      dispatch(addArticles(result.articles));
    } catch (error) {
      console.log(error);
    }
  }
}