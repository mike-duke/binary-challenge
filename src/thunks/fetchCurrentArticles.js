import { addCurrentArticles, addErrorMessage } from '../actions';

export const fetchCurrentArticles = (url) => {
  return async (dispatch) => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      } else if (response.ok && response.totalResults <= 0) {
        dispatch(addErrorMessage('There are no articles that match this query'))
      }
      const result = await response.json();
      dispatch(addCurrentArticles(result.articles));
    } catch (error) {
      dispatch(addErrorMessage(error.message));
    }
  }
}