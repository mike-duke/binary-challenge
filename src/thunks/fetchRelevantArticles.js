import { addRelevantArticles, addErrorMessage, isLoading } from '../actions';

export const fetchRelevantArticles = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      } else if (response.ok && response.totalResults <= 0) {
        throw Error('There are no articles that match this query');
      }
      dispatch(isLoading(false));
      const result = await response.json();
      dispatch(addRelevantArticles(result.articles));
    } catch (error) {
      dispatch(addErrorMessage(error.message))
    }
  }
}