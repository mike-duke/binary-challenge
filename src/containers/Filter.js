import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTopic } from '../actions';
import { fetchRelevantArticles } from '../thunks/fetchRelevantArticles';
import apiKey from '../apiKey';
import uuid from 'uuid';

export class Filter extends Component {
  
  handleChange = (event) => {
    const { name, value } = event.target;
    const lowerCaseValue = value.toLowerCase();
    let url;
    let selection;
    if (name === 'filter-select') {
      selection = event.target.value;
      url = `https://newsapi.org/v2/everything?q=+(parent OR parents) AND +(kid OR kids OR child OR children) AND ${selection}&pageSize=20&language=en&sortBy=popularity&apiKey=${apiKey}`;
    }
    
    if (name === 'source-select' && value.includes('.')) {
      selection = this.props.topic;
      url = `https://newsapi.org/v2/everything?q=+(parent OR parents) AND +(kid OR kids OR child OR children) AND ${selection}&domains=${lowerCaseValue}&language=en&sortBy=popularity&apiKey=${apiKey}`;
    } else if (name === 'source-select' && !value.includes('.')) {
      selection = this.props.topic;
      url = `https://newsapi.org/v2/everything?q=+(parent OR parents) AND +(kid OR kids OR child OR children) AND ${selection}&sources=${lowerCaseValue}&language=en&sortBy=popularity&apiKey=${apiKey}`;
    }
    this.props.addRelevantArticlesToStore(url);
    this.props.addTopicToStore(selection);
  }

  filterSelections = () => {
    const namesArray = this.props.relevantArticles.reduce((names, article) => {
      if (!names.includes(article.source.name)) {
        names.push(article.source.name);
      }
      return names;
    }, []);
    
    const sourcesObj = this.props.relevantArticles.reduce((sources, article) => {
      sources[article.source.name] = article.source.id;
      return sources;
    }, {});

    return namesArray.map((name) => {
      if (name.includes('.')) {
        return <option key={uuid()} value={name}>{name}</option>
      } else {
        return <option key={uuid()} value={sourcesObj[name].id}>{name}</option>
      }
    })
  }

  render() {
    const sourceOptions = this.filterSelections();

    return (
      <div className="filter">
        <select name="filter-select" id="filter-select" onChange={this.handleChange}>
          <option value="">Select a different topic to gather articles</option>
          <option value="anxiety">Anxiety</option>
          <option value="depression">Depression</option>
          <option value="ADHD">ADHD</option>
          <option value="autism">Autism</option>
        </select>

        <select name="source-select" id="source-select" onChange={this.handleChange}>
          <option value="">Select a source to gather specific articles</option>
          {sourceOptions}
        </select>
        <h2>Torcano</h2>
    </div>
  )
}
}

export const mapStateToProps = (state) => ({
  relevantArticles: state.relevantArticles,
  currentArticles: state.currentArticles,
  topic: state.topic
});

export const mapDispatchToProps = (dispatch) => ({
  addTopicToStore: (selection) => dispatch(addTopic(selection)),
  addRelevantArticlesToStore: (url) => dispatch(fetchRelevantArticles(url))
});

export default connect(mapStateToProps, mapDispatchToProps)(Filter);

Filter.propTypes = {
  relevantArticles: PropTypes.array,
  currentArticles: PropTypes.array,
  topic: PropTypes.string,
  addTopicToStore: PropTypes.func,
  addRelevantArticlesToStore: PropTypes.func
}