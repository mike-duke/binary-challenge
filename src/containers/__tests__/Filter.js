import React from 'react';
import { shallow } from 'enzyme';
import { Filter, mapStateToProps, mapDispatchToProps } from '../Filter';
import { addTopic } from '../../actions';
import { fetchRelevantArticles } from '../../thunks/fetchRelevantArticles';

jest.mock('../../thunks/fetchRelevantArticles');

describe('Filter', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Filter relevantArticles={[]} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('should return a props object with the correct properties', () => {
      const mockState = {
        relevantArticles: [],
        currentArticles: [],
        topic: 'Bacon'
      }

      const expected = {
        relevantArticles: [],
        currentArticles: [],
        topic: 'Bacon'
      }

      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);

    it('should dispatch addTopic when addTopicToStore is called', () => {
      const expected = addTopic('Burgers');

      mappedProps.addTopicToStore('Burgers');
      expect(mockDispatch).toHaveBeenCalledWith(expected);
    });

    it('should dispatch fetchRelevantArticles thunk when addRelevantArticles is called', () => {
      const url = 'somewhere.com';
      const expected = fetchRelevantArticles(url);

      mappedProps.addRelevantArticlesToStore(url);
      expect(mockDispatch).toHaveBeenCalledWith(expected);
    })
  });
});