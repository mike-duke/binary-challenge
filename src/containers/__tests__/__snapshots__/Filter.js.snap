// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Filter
    relevantArticles={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <select
          id="filter-select"
          name="filter-select"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Select a different topic to gather articles
          </option>
          <option
            value="anxiety"
          >
            Anxiety
          </option>
          <option
            value="depression"
          >
            Depression
          </option>
          <option
            value="ADHD"
          >
            ADHD
          </option>
          <option
            value="autism"
          >
            Autism
          </option>
        </select>,
        <select
          id="source-select"
          name="source-select"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Select a source to gather specific articles
          </option>
        </select>,
        <h2>
          Torcano
        </h2>,
      ],
      "className": "filter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value=""
            >
              Select a different topic to gather articles
            </option>,
            <option
              value="anxiety"
            >
              Anxiety
            </option>,
            <option
              value="depression"
            >
              Depression
            </option>,
            <option
              value="ADHD"
            >
              ADHD
            </option>,
            <option
              value="autism"
            >
              Autism
            </option>,
          ],
          "id": "filter-select",
          "name": "filter-select",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select a different topic to gather articles",
              "value": "",
            },
            "ref": null,
            "rendered": "Select a different topic to gather articles",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Anxiety",
              "value": "anxiety",
            },
            "ref": null,
            "rendered": "Anxiety",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Depression",
              "value": "depression",
            },
            "ref": null,
            "rendered": "Depression",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "ADHD",
              "value": "ADHD",
            },
            "ref": null,
            "rendered": "ADHD",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Autism",
              "value": "autism",
            },
            "ref": null,
            "rendered": "Autism",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value=""
            >
              Select a source to gather specific articles
            </option>,
            Array [],
          ],
          "id": "source-select",
          "name": "source-select",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select a source to gather specific articles",
              "value": "",
            },
            "ref": null,
            "rendered": "Select a source to gather specific articles",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Torcano",
        },
        "ref": null,
        "rendered": "Torcano",
        "type": "h2",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <select
            id="filter-select"
            name="filter-select"
            onChange={[Function]}
          >
            <option
              value=""
            >
              Select a different topic to gather articles
            </option>
            <option
              value="anxiety"
            >
              Anxiety
            </option>
            <option
              value="depression"
            >
              Depression
            </option>
            <option
              value="ADHD"
            >
              ADHD
            </option>
            <option
              value="autism"
            >
              Autism
            </option>
          </select>,
          <select
            id="source-select"
            name="source-select"
            onChange={[Function]}
          >
            <option
              value=""
            >
              Select a source to gather specific articles
            </option>
          </select>,
          <h2>
            Torcano
          </h2>,
        ],
        "className": "filter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value=""
              >
                Select a different topic to gather articles
              </option>,
              <option
                value="anxiety"
              >
                Anxiety
              </option>,
              <option
                value="depression"
              >
                Depression
              </option>,
              <option
                value="ADHD"
              >
                ADHD
              </option>,
              <option
                value="autism"
              >
                Autism
              </option>,
            ],
            "id": "filter-select",
            "name": "filter-select",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select a different topic to gather articles",
                "value": "",
              },
              "ref": null,
              "rendered": "Select a different topic to gather articles",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Anxiety",
                "value": "anxiety",
              },
              "ref": null,
              "rendered": "Anxiety",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Depression",
                "value": "depression",
              },
              "ref": null,
              "rendered": "Depression",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ADHD",
                "value": "ADHD",
              },
              "ref": null,
              "rendered": "ADHD",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Autism",
                "value": "autism",
              },
              "ref": null,
              "rendered": "Autism",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value=""
              >
                Select a source to gather specific articles
              </option>,
              Array [],
            ],
            "id": "source-select",
            "name": "source-select",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select a source to gather specific articles",
                "value": "",
              },
              "ref": null,
              "rendered": "Select a source to gather specific articles",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Torcano",
          },
          "ref": null,
          "rendered": "Torcano",
          "type": "h2",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
